### DISCUSSION TITLE
Dimmers PSM / PWM

### DISCUSSION THREAD

**Discussion by threedom** [theyosh/TerrariumPI#521] [Q&A]
Really looking forward to trying this out with my terrarium.
Did anyone try using this with a dimmer to automatically adjust the intensity of heating lamps? If so, is a PSM or PWM dimmer recommended for a 75w infrared lamp?

Thank you in advance!

**Comment by escomputers** [comment]
Hello, I personally used a PWM dimmer. It works very well. Here's a good tutorial for setup.
(https://medium.com/@jinky32/connecting-a-pulse-width-modulation-pwm-dimmer-switch-board-to-a-raspberry-pi-for-use-with-a16866df813b)

**Reply by threedom** [reply]
Thank you for your reply. I actually got this dimmer https://www.amazon.com/gp/product/B06Y1GVG26/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1#customerReviews because I wanted to control both a lamp and a heat mat with one device. 
I connected it to the Pi and conceptually it works (in TerrariumPi I picked the Mosfet Dimmer preset since this was the only one that actually made it dimm the light, the PWM preset would only switch the light on and off). 
However, the light is flickering a lot, especially at lower voltages. Did anyone else have these issues and could potentially help?

(Sorry for the beginner question, but Google and forum searches did not yield any results that helped.)

**Reply by theyosh** [reply]
Have a look at: https://github.com/theyosh/TerrariumPI/issues/486#issuecomment-751502614

**Reply by threedom** [reply]
Appreciate the comment but I had already tried that.
When I change the frequency setting the following behaviors are triggered:

- <110 flickering
- at 119 smoothly drops from max brighness to almost 0 in about 1 second (in a loop)
- at 120 brightness increases smoothly from off to 100% and then abruptly switches off (speed of this process depends on what percentage is selected in TerrariumPI)
- >120 (e.g. 140) flickering
- at 240 brightness increases smoothly from low to 100% and then drops back to low
- very high (e.g. 720, 24k) only switches on to full brightness (sometimes only at high percentage in TerrariumPi


Long story short:
It looks like somehow the dimmer should be able to just maintain a lower brightness since it can smoothly transition the bulb through brightness levels. But somehow it doesn't maintain one brightness level.

I am also happy to create a post in the Issues. I was just not sure this qualifies.

**Reply by theyosh** [reply]
I am afraid that for now the only solution is to play with the DIMMER_FREQ and DIMMER_MAXDIM values.

The new code works differently, and hopefully better. For me it is impossible to test all the available dimmers and light combinations. Some lights work better with dimmers then others. So it is a bit of trial and error. But not something that I can fix now in the code.

**Reply by threedom** [reply]
Fully understand, obviously not something that is your responsibility. I really like what you have created. I might post in issues at some point to see if someone else has the same challenge and found a solution.
